File: 1756-minimum-deletions-to-make-string-balanced/minimum-deletions-to-make-string-balanced.cpp
class Solution {
public:
    int minimumDeletions(string s) {
        ios_base::sync_with_stdio(false);
        cin.tie(nullptr);

        s += 'b';
        int n = s.size(), cnt = 0, result = 1e9;
        vector<int> count(n);
        
        for (int i = n - 1; i >= 0; i--) {
            cnt += s[i] == 'a' ? 1 : 0;
            count[i] = cnt;
        }
        
        cnt = 0;
        for (int i = 0; i < n; i++) {
            result = min(result, cnt + count[i]);
            cnt += s[i] == 'b' ? 1 : 0;
        }
        
        return result;
    }
};

File: 1756-minimum-deletions-to-make-string-balanced/minimum-deletions-to-make-string-balanced.rs
impl Solution {
    pub fn minimum_deletions(s: String) -> i32 {
        let s = s.as_bytes();
        let n = s.len();
        let mut cnt = 0;
        let mut res = i32::MAX;
        let mut right = vec![0; n + 1];

        for i in (0..n).rev() {
            if s[i] == b'a' { cnt += 1; }
            right[i] = cnt;
        }

        cnt = 0;
        for i in 0..n {
            res = res.min(cnt + right[i + 1]);
            if s[i] == b'b' { cnt += 1; }
        }

        res
    }
}